vendor/bam/src/json/frame.rs

```diff
@@ -1,191 +1,29 @@
-        frame: json::Frame,
-    ) -> Result<Option<Box<dyn Future<Item = json::Frame, Error = ()> + Send + 'static>>, Error>
-    {
-        match frame.frame_type.as_str() {
-            "REQUEST" => {
-                if frame.id < self.next_expected_id {
-                    return Err(Error::OutOfOrderRequest);
-                }
-
-                self.next_expected_id = frame.id + 1;
-
-                let frame_id = frame.id;
-
-                let response_future = serde_json::from_value(frame.payload)
-                    .map_err(malformed_request)
-                    .and_then(|request| self.validate_request(request))
-                    .and_then(|request| self.dispatch_request(request))
-                    .unwrap_or_else(|error_response| Box::new(future::ok(error_response)))
-                    .and_then(move |response| Ok(response.into_frame(frame_id)));
-
-                Ok(Some(Box::new(response_future)))
-            }
-            "RESPONSE" => {
-                let mut response_source = self.response_source.lock().unwrap();
-
-                let sender = response_source
-                    .get_awaiting_response(frame.id)
-                    .ok_or(Error::UnexpectedResponse)?;
-
-                log::debug!(
-                    "attempting to deserialize payload '{:?}' of frame {} as RESPONSE",
-                    frame.payload,
-                    frame.id
-                );
-
-                let response = serde_json::from_value(frame.payload);
-
-                match response {
-                    Ok(response) => {
-                        log::debug!("dispatching {:?} to stored handler", response);
-                        sender.send(response).unwrap()
-                    }
-                    Err(e) => log::error!(
-                        "payload of frame {} is not a well-formed RESPONSE: {:?}",
-                        frame.id,
-                        e
-                    ),
-                }
-
-                Ok(None)
-            }
-            _ => Err(Error::UnknownFrameType(frame.frame_type)),
-        }
-    }
-}
-
-fn malformed_request(error: serde_json::Error) -> json::Response {
-    log::warn!("incoming request was malformed: {:?}", error);
-
-    json::Response::new(Status::SE(0))
-}
-
-fn unknown_request_type(request_type: &str) -> json::Response {
-    log::warn!("request type '{}' is unknown", request_type);
-
-    json::Response::new(Status::SE(2))
-}
+#[derive(Deserialize, Serialize, PartialEq, Debug)]
+#[serde(rename_all = "UPPERCASE")]
+pub enum FrameType {
+    Request,
+    Response,
 
-fn unknown_mandatory_headers(unknown_headers: UnknownMandatoryHeaders) -> json::Response {
-    json::Response::new(Status::SE(1)).with_header(
-        "Unsupported-Headers",
-        Header::with_value(unknown_headers)
-            .expect("list of strings should serialize to serde_json::Value"),
-    )
+    // Too lazy ATM to write a proper deserializer that preserves the value 🙃
```

Comment:
:upside_down_face: TODO?

Reply:
Yeah, donno :joy: 

It is going to be revised soon so I didn't bother :man_shrugging: